aten/src/ATen/cuda/CUDASparseDescriptors.cpp

```diff
@@ -0,0 +1,168 @@
+#include <ATen/native/cuda/MiscUtils.h>
+
+#if AT_USE_CUSPARSE_GENERIC_API()
+
+namespace at {
+namespace cuda {
+namespace sparse {
+
+namespace {
+
+// If a specific GPU model does not provide native support for a given data
+// type, cuSparse routines return CUSPARSE_STATUS_ARCH_MISMATCH error
+void check_supported_cuda_type(cudaDataType cuda_type) {
+  if (cuda_type == CUDA_R_16F) {
+    cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+    TORCH_CHECK(
+        prop->major >= 5 && ((10 * prop->major + prop->minor) >= 53),
+        "Sparse operations with CUDA tensors of Float16 type are not supported on GPUs with compute capability < 5.3 (current: ",
+        prop->major,
+        ".",
+        prop->minor,
+        ")");
+  }
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11000
+  if (cuda_type == CUDA_R_16BF) {
+    cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+    TORCH_CHECK(
+        prop->major >= 8,
+        "Sparse operations with CUDA tensors of BFloat16 type are not supported on GPUs with compute capability < 8.0 (current: ",
+        prop->major,
+        ".",
+        prop->minor,
+        ")");
+  }
+#endif
+}
+
+} // anonymous namespace
+
+cusparseIndexType_t getCuSparseIndexType(const c10::ScalarType& scalar_type) {
+  if (scalar_type == c10::ScalarType::Int) {
+    return CUSPARSE_INDEX_32I;
+  } else if (scalar_type == c10::ScalarType::Long) {
+    return CUSPARSE_INDEX_64I;
+  } else {
+    TORCH_INTERNAL_ASSERT(
+        false, "Cannot convert type ", scalar_type, " to cusparseIndexType.");
+  }
+}
+
+CuSparseDnMatDescriptor::CuSparseDnMatDescriptor(const Tensor& input) {
+  TORCH_INTERNAL_ASSERT(input.layout() == kStrided);
+  IntArrayRef input_strides = input.strides();
+  IntArrayRef input_sizes = input.sizes();
+  auto ndim = input.dim();
+  TORCH_INTERNAL_ASSERT(ndim >= 2);
+  auto rows = input_sizes[ndim - 2];
+  auto cols = input_sizes[ndim - 1];
+
+  bool is_column_major =
+      at::native::is_blas_compatible_column_major_order(input);
+  bool is_row_major = at::native::is_blas_compatible_row_major_order(input);
+  TORCH_INTERNAL_ASSERT(
+      is_column_major || is_row_major,
+      "Expected either row or column major contiguous input.");
+
+  auto leading_dimension =
+      is_row_major ? input_strides[ndim - 2] : input_strides[ndim - 1];
+  auto order = is_row_major ? CUSPARSE_ORDER_ROW : CUSPARSE_ORDER_COL;
+
+  void* values_ptr = input.data_ptr();
+
+  cudaDataType value_type = ScalarTypeToCudaDataType(input.scalar_type());
+  check_supported_cuda_type(value_type);
+
+  cusparseDnMatDescr_t raw_descriptor;
+  TORCH_CUDASPARSE_CHECK(cusparseCreateDnMat(
+      &raw_descriptor,
+      rows,
+      cols,
+      leading_dimension,
+      values_ptr,
+      value_type,
+      order));
+
+  if (ndim > 2) {
+    auto batch_count =
+        at::native::cuda_int_cast(at::native::batchCount(input), "batch_count");
+    auto batch_stride = input_strides[ndim - 3];
```

Comment:
Since we can have arbitrary ndim here, why do we choose input_strides[ndim - 3] as the batch_stride?

Reply:
When matrices are stored in the last two dimensions in order to iterate over all matrices in n-dimensional tensor we need to use `input_strides[ndim - 3]`.
`input_strides[ndim - 1]` is used to iterate over scalar,
`input_strides[ndim - 2]` is used to iterate over vectors,
`input_strides[ndim - 3]` is used to iterate over matrices,
```
for (int64_t batch_id = 0; batch_id < prod(tensor.shape[:-2]); batch_id++) {
    // in contiguous case tensor.stride(-3) == tensor.stride(-1)*tensor.stride(-2)
    current_matrix_ptr = tensor_data[batch_id*tensor.stride(-1)*tensor.stride(-2)]
}
```

Reply:
Sure, but that'd be for a 3-dim Tensor only, no? But we're checking for ndim greater than 2, so what about the 4d case?

Reply:
No, it would be N-dim Tensor, where N >= 3.
The data of tensors is a 1D array in memory and elements are accessed using a single `operator[]`. The tensors can be 543d and the code will remain correct, last two dimensions are matrices and all other could as well be reshaped to a 3d tensor for iteration.
```python
In [1]: import torch

In [2]: a = torch.randn(2, 2, 2, 2) # 4d case

In [3]: a
Out[3]:
tensor([[[[-1.0814, -0.5789],
          [ 1.0179, -2.3714]],

         [[ 0.2434,  0.4329],
          [ 0.4716, -0.7596]]],


        [[[-0.4049,  1.5974],
          [-0.9816,  1.1460]],

         [[-0.1591,  1.3282],
          [-0.0039,  0.1529]]]])

In [4]: a.flatten().shape
Out[4]: torch.Size([16])

In [5]: for batch_id in range(2*2):
   ...:     batch_stride = a.stride(-3)
   ...:     current_matrix_start = a.flatten()[batch_id * batch_stride]
   ...:     print(current_matrix_start)
   ...:
tensor(-1.0814)
tensor(0.2434)
tensor(-0.4049)
tensor(-0.1591)
```

Reply:
Sure, but this only is true if the memory is contiguous. I guess we have a contiguity check somewhere. In any case we're only supporting 2d for now.